File: app/src/main/java/com/example/githubrepo/services/event/BusEvent.java
package com.example.githubrepo.services.event;

import java.util.Collections;

/**
 * Created by sarah_neo on 22/02/2017.
 */

public abstract class BusEvent<T> {
    public enum EventType {
        REPOS;

        public BusEvent getEmptyEvent() {
            switch (this) {
                case REPOS:
                    return new LoadReposEvent().setData(Collections.EMPTY_LIST);
                default:
                    throw new IllegalArgumentException("No such event registered");
            }
        }
    }

    public abstract BusEvent setData(T data);

    public abstract T getData();

    public abstract EventType getType();

    public String getName() {
        return getType().name();
    }
}

File: app/src/main/java/com/example/githubrepo/services/event/BusEvent.kt
package com.example.githubrepo.services.event

abstract class BusEvent<T> {
    enum class EventType {
        REPOS;

        val emptyEvent: BusEvent<*>
            get() = when (this) {
                REPOS -> LoadReposEvent().setData(null)
            }
    }

    abstract fun setData(data: T): BusEvent<*>?
    abstract val data: T
    abstract val type: EventType
    val name: String
        get() = type.name
}